x-pack/libbeat/management/manager.go

```diff
@@ -0,0 +1,197 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package management
+
+import (
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/libbeat/common/reload"
+
+	"github.com/satori/go.uuid"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/common/cfgwarn"
+	"github.com/elastic/beats/libbeat/feature"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/x-pack/libbeat/management/api"
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/management"
+)
+
+func init() {
+	management.Register("x-pack", NewConfigManager, feature.Beta)
+}
+
+// ConfigManager handles internal config updates. By retrieving
+// new configs from Kibana and applying them to the Beat
+type ConfigManager struct {
+	config   *Config
+	logger   *logp.Logger
+	client   *api.Client
+	beatUUID uuid.UUID
+	done     chan struct{}
+	registry *reload.Registry
+	wg       sync.WaitGroup
+}
+
+// NewConfigManager returns a X-Pack Beats Central Management manager
+func NewConfigManager(registry *reload.Registry, beatUUID uuid.UUID) (management.ConfigManager, error) {
+	c := defaultConfig()
+	if err := c.Load(); err != nil {
+		return nil, errors.Wrap(err, "reading central management internal settings")
+	}
+	return NewConfigManagerWithConfig(c, registry, beatUUID)
+}
+
+// NewConfigManagerWithConfig returns a X-Pack Beats Central Management manager
+func NewConfigManagerWithConfig(c *Config, registry *reload.Registry, beatUUID uuid.UUID) (management.ConfigManager, error) {
+	var client *api.Client
+	if c.Enabled {
+		var err error
+
+		// Ignore kibana version to avoid permission errors
+		c.Kibana.IgnoreVersion = true
+
+		client, err = api.NewClient(c.Kibana)
+		if err != nil {
+			return nil, errors.Wrap(err, "initializing kibana client")
+		}
+	}
+
+	return &ConfigManager{
+		config:   c,
+		logger:   logp.NewLogger(management.DebugK),
+		client:   client,
+		done:     make(chan struct{}),
+		beatUUID: beatUUID,
+		registry: registry,
+	}, nil
+}
+
+// Enabled returns true if config management is enabled
+func (cm *ConfigManager) Enabled() bool {
+	return cm.config.Enabled
+}
+
+// Start the config manager
+func (cm *ConfigManager) Start() {
```

Comment:
It's helpful to call methods and functions which launches new goroutines on their bodies with a specify suffix like `StartAsync` or `StartGoroutine`. This way, future code readers won't assume synchronous work

Reply:
I was also thinking that it's not a bad idea to leave it with `Start` but make it a block call. The effect is similar to the one you get with `http.ListenAndServe` and the caller gets also a clear idea that he/she needs have some dancing gouroutines after this call (because he/she will use `go Start()` in the code)

Reply:
This implements our Runner interface, with Start/Stop methods. In  general I think we follow this style: `Start` is non blocking, sometimes we use `Run` as it's blocking counterpart